core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -626,7 +642,7 @@ object IO {
     override def tag = Tags.Uninterruptible
   }
 
-  final class Sleep private[IO] (val duration: Duration) extends IO[Nothing, Unit] {
+  final class Sleep private[IO] (val duration: Duration.Finite) extends IO[Nothing, Unit] {
```

Comment:
Change this to accept infinite duration, and simply never schedule the action if the duration is infinite (`IO.sleep(Duration.Infinite)` should be the same as `never`).